File: java/devtools/src/main/java/org/pactman/nonprofitcheckplus/devtools/ApiDate.java
package org.pactman.nonprofitcheckplus.devtools;

import java.time.Instant;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.time.format.DateTimeFormatter;
import java.time.temporal.ChronoUnit;
import java.util.Locale;

/**
 * The date format the API uses, and the arithmetic the freshness examples need.
 *
 * <p>The API formats every date as {@code M/dd/yyyy h:mm:ss a}. It is a display
 * format, not ISO 8601, so anything that parses it is guessing at a locale and a
 * time zone — which is exactly why the SDK hands these fields back as strings
 * and leaves the decision here, where an example can be explicit about it.
 */
public final class ApiDate {

    private static final DateTimeFormatter FORMAT =
            DateTimeFormatter.ofPattern("M/dd/yyyy h:mm:ss a", Locale.US);

    private ApiDate() {
    }

    /**
     * A date this many days in the past, formatted the way the API formats one.
     *
     * <p>Fixture dates are generated relative to today so the freshness examples
     * stay meaningful however long after they were written they are run.
     *
     * @param daysAgo how many days back.
     * @return the formatted date.
     */
    public static String daysAgo(long daysAgo) {
        return FORMAT.format(LocalDateTime.now().minusDays(daysAgo));
    }

    /**
     * How many days ago a date the API returned was.
     *
     * @param apiDate the date as the API formatted it, or {@code null}.
     * @return the age in days, or {@code null} when the value is absent or
     *         cannot be read as a date.
     */
    public static Long ageInDays(String apiDate) {
        if (apiDate == null || apiDate.trim().isEmpty()) {
            return null;
        }

        try {
            LocalDateTime when = LocalDateTime.parse(apiDate.trim(), FORMAT);

            return ChronoUnit.DAYS.between(
                    when.atZone(ZoneId.systemDefault()).toInstant(), Instant.now());
        } catch (java.time.format.DateTimeParseException unparseable) {
            // A format the API changed, or a value that was never a date. The
            // caller decides what that means; guessing here would be worse.
            return null;
        }
    }
}

File: go/examples/internal/support/apidate.go
package support

import (
	"strings"
	"time"
)

// The API formats timestamps for display, not for machines: M/DD/YYYY h:mm:ss AM.
//
// Parsing them means assuming a locale and a time zone, which is why the SDK
// hands them back as strings and leaves the assumption to code like this, where
// it is visible and can be argued with. The variants below are accepted because
// a display format is not a contract.
var apiDateLayouts = []string{
	"1/02/2006 3:04:05 PM",
	"1/2/2006 3:04:05 PM",
	"01/02/2006 3:04:05 PM",
	"1/02/2006 15:04:05",
	"2006-01-02",
}

// ParseAPIDate reads a timestamp the API returned. The second result is false
// when the value is not in a form this reader recognizes.
func ParseAPIDate(value string) (time.Time, bool) {
	trimmed := strings.TrimSpace(value)
	if trimmed == "" {
		return time.Time{}, false
	}

	for _, layout := range apiDateLayouts {
		if parsed, err := time.ParseInLocation(layout, trimmed, time.Local); err == nil {
			return parsed, true
		}
	}

	return time.Time{}, false
}

// AgeInDays returns how many days ago a timestamp was, or nil when the API
// returned no value or one this reader cannot parse.
//
// An unparseable date returns nil rather than zero: "today" is a much worse
// answer than "I do not know".
func AgeInDays(value *string) *int64 {
	if value == nil {
		return nil
	}

	parsed, ok := ParseAPIDate(*value)
	if !ok {
		return nil
	}

	days := int64(time.Since(parsed) / (24 * time.Hour))

	return &days
}
